homeassistant/components/voice_assistant/pipeline.py

```diff
@@ -56,69 +51,165 @@ class Pipeline:
+            language=self.language,
+            agent_id=self.pipeline.conversation_engine,
         )
 
-        event_callback(
+        self.event_callback(
             PipelineEvent(
                 PipelineEventType.INTENT_FINISH,
                 {"intent_output": conversation_result.as_dict()},
             )
         )
 
-        event_callback(
+        return conversation_result
+
+    async def text_to_speech(self, tts_input: str) -> str:
+        """Run text to speech portion of pipeline. Returns URL of TTS audio."""
+        self.event_callback(
             PipelineEvent(
-                PipelineEventType.RUN_FINISH,
+                PipelineEventType.TTS_START,
+                {
+                    "engine": self.pipeline.tts_engine,
+                    "tts_input": tts_input,
+                },
+            )
+        )
+
+        manager: tts.SpeechManager = self.hass.data[tts.DOMAIN]
+        tts_url = await manager.async_get_url_path(self.pipeline.tts_engine, tts_input)
+
+        self.event_callback(
+            PipelineEvent(
+                PipelineEventType.TTS_FINISH,
+                {"tts_output": tts_url},
             )
         )
+
+        return tts_url
+
+
+@dataclass
+class PipelineRequest(ABC):
+    """Request to for a pipeline run."""
+
+    @abstractmethod
+    async def execute(
+        self, run: PipelineRun, timeout: int | float | None = DEFAULT_TIMEOUT
+    ):
+        """Run pipeline with request info and context."""
+
+
+@dataclass
+class TextPipelineRequest(PipelineRequest):
+    """Request to run the text portion only of a pipeline."""
+
+    intent_input: str
+    conversation_id: str | None = None
+
+    async def execute(
+        self, run: PipelineRun, timeout: int | float | None = DEFAULT_TIMEOUT
+    ):
+        """Run text portion of pipeline."""
+        await asyncio.wait_for(
+            self._execute(run),
+            timeout=timeout,
+        )
+
+    async def _execute(
+        self,
+        run: PipelineRun,
+    ):
+        await run.start()
+        await run.recognize_intent(self.intent_input, self.conversation_id)
+        await run.finish()
+
+
+@dataclass
+class AudioPipelineRequest(PipelineRequest):
+    """Request to full pipeline from audio input (stt) to audio output (tts)."""
+
+    intent_input: str  # this will be changed to stt audio
+    conversation_id: str | None = None
+
+    async def execute(
+        self, run: PipelineRun, timeout: int | float | None = DEFAULT_TIMEOUT
+    ):
+        """Run text portion of pipeline."""
+        await asyncio.wait_for(
+            self._execute(run),
+            timeout=timeout,
+        )
+
+    async def _execute(self, run: PipelineRun):
+        await run.start()
+
+        # stt will go here
+
+        conversation_result = await run.recognize_intent(
+            self.intent_input, self.conversation_id
+        )
+
+        tts_input = conversation_result.response.speech.get("plain", {}).get(
```

Comment:
Would it make sense to have the voice pipeline extend the text pipeline and call this part with super() ? Or are we making things too complicated.

Reply:
I considered that, but I thought not having `intent_input` on the `AudioPipelineRequest` would be a good thing since it differentiates the two requests well.